### PR TITLE
The Great Glob Import Extinction (and requisite finishing touches)

### LINKED ISSUE #101: Meta-issue for ergonomic advancements
Followup to #90.  Here's the grander scheme as I see it:

* [x] Add inherent methods
  * [x] To HLists and Coproducts (done in #99)
  * [x] Anything else? (checked while doing #103)
  * [ ] For HMappable and friends now that the impls `for &Type` are gone
* [x] Reexport the most important things at the root (done in #103)
  * HCons, HNil, Coproduct, HList
  * What else from the other modules?
* [x] Introduce a prelude module for traits that need to be imported to be used (done in #103)
  * extension traits for types outside frunk
  * things like `HMappable` that have impls for `&HCons`
    * Or we can take this opportunity to introduce a more general `Mappable` trait
* ~~[#] **(maybe!)** Rename the traits in line with the rust convention of `Verb::verb`~~
  * I don't think this is worth it right now
* Take advantage of the new features in docs and examples
  * use turbofish!
  * no more `use frunk::hlist::*`
  * [x] in `README.md`
  * [ ] in examples
    * if necessary, add `frunk` as dev dependency to frunk-core so doc examples can use frunk's prelude (or add a prelude to frunk-core, but I think having two makes it harder to maintain)

### PR DESCRIPTION
**Pull request by ExpHP**
As before, I tried to break this up into individually readible commits.

It includes several of the last-minute breaking changes [mentioned here](https://github.com/lloydmeta/frunk/issues/101#issuecomment-379569647), each as its own commit (or two) near the beginning. This is followed by one really big commit that corrects all of the examples and tests to use the new use paths.

Closes #101

### CODE REVIEW

**Comment by ExpHP** [comment]
wtf why is 'doc updates for Func' listed after TGGIE, I thought I fixed all of the timestamps >_<

whatever, at least that's the only one out of order this time.

**Comment by lloydmeta** [comment]
My simple mind is saying just a `poly.rs` module in frunk_core might work? Open to other ideas of course :)

File: core/src/indices.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+use std::marker::PhantomData;
+
+// Largely lifted from https://github.com/Sgeo/hlist/blob/master/src/lib.rs#L30
+
+/// Used as an index into an `HList`.
+///
+/// `Here` is 0, pointing to the head of the HList.
+///
+/// Users should normally allow type inference to create this type.
+pub struct Here(());
```
Comment (Centril) [inline core/src/indices.rs]: Should probably use `{ priv: () }` here to make the privacy thingy more intentional-looking.

**Comment by ExpHP** [comment]
I considered it and am open to changing it, but note that there is an
observable difference to the user between a regular struct with a private
member and a tuple struct with a private member. (the trouble is, I'm not
actually sure at this point which is better for reified indices, or if it
even matters at all!)

**Comment by ExpHP** [comment]
Scratch that, actually. Even tuple structs can use struct expression syntax
and struct pattern syntax, so I guess regular structs are the clear winner
for maximum flexibility in the future. I will change both Here and There.

**Comment by ExpHP** [comment]
Oops, mail replies don't appear where I thought they would on this page. =P

**Comment by ExpHP** [comment]
> My simple mind is saying just a poly.rs module in frunk_core might work? Open to other ideas of course :)

This is definitely possible, although I don't think `ToRef` will quite fit in there.  My trouble is with having too many little modules for small groups of things; I feel this harms discoverability.

I tried making them live at the root but I didn't like how the resulting documentation for `frunk_core/index.html` looked (it made them look too important); so for now I've moved them into a public module called `traits`.  This name feels vague enough to be able to cover most of the things I want to remove from HList; however, it goes counter to the naming scheme of the rest of frunk (which mostly groups things by feature).

File: core/src/hlist.rs [hunk 2]
```diff
@@ -331,7 +331,7 @@ macro_rules! gen_inherent_methods {
             /// * A single closure (for mapping an HList that is homogenous).
             /// * A single [`Poly`].
             ///
-            /// [`Poly`]: struct.Poly.html
+            /// [`Poly`]: ../traits/struct.Poly.html
```
Comment (lloydmeta) [inline core/src/hlist.rs]: Marginally unrelated question: are these checked by the doc-generator?

Comment (ExpHP) [reply core/src/hlist.rs]: I installed linkchecker from `rustc/src/tools` and use it religiously

Comment (ExpHP) [reply core/src/hlist.rs]: To clarify, I think the process is:

```rust
git clone https://github.com/rust-lang/rust
cd rust 
git submodule update # or init, I forget
cd src/tools/linkchecker
cargo install
```

Comment (Centril) [reply core/src/hlist.rs]: @ExpHP Oh; I feel like I just discovered ice-cream. Marvelous!

**Pull request merged**